src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs

```diff
@@ -1264,28 +1283,139 @@ public static bool IsPatternContext(this SyntaxTree syntaxTree, SyntaxToken left
+                    return true;
+                }
+
+                // e is not $$
+                if (node.IsKind(SyntaxKind.NotPattern))
+                {
+                    return true;
+                }
+
+                // e is ($$ 1 or 2)
+                if (node.IsKind(SyntaxKind.ParenthesizedPattern))
+                {
+                    return true;
+                }
+            }
+#endif
+
+            return false;
+        }
+
+        public static bool IsAtEndOfPattern(this SyntaxTree syntaxTree, SyntaxToken leftToken, int position)
+        {
+#if !CODE_STYLE
+            leftToken = leftToken.GetPreviousTokenIfTouchingWord(position);
+
+            // This check should cover the majority of cases, e.g.:
+            // e is 1 $$
+            // e is >= 0 $$
+            // e is { P: (1 $$
+            if (leftToken.IsLastTokenOfNode<PatternSyntax>(out var patternSyntax, includeZeroWidth: false))
+            {
+                // Patterns such as 'e is not $$', 'e is 1 or $$', and 'e is ($$' should be invalid here.
+                if (!(patternSyntax is UnaryPatternSyntax || patternSyntax is BinaryPatternSyntax || patternSyntax is RecursivePatternSyntax))
+                {
+                    return true;
+                }
+            }
+
+            // e is C.P
+            // e is int $$
+            if (leftToken.IsLastTokenOfNode<BinaryExpressionSyntax>(out var binaryExpression, includeZeroWidth: false))
+            {
+                if (binaryExpression.OperatorToken.IsKind(SyntaxKind.IsKeyword) && !binaryExpression.Right.Span.IsEmpty)
+                {
+                    return true;
+                }
+            }
+
+            var node = leftToken.Parent;
+
+            // e is int $$
+            // The parser can parse the above example as both a BinaryExpressionSyntax and DeclarationPatternSyntax,
+            // We already check for BinaryExpressionSyntax above, so now we just need to check for DeclarationPatternSyntax.
+            // Normally, checking for DeclarationPatternSyntax would be handled by the first check of the method,
+            // (i.e. leftToken.IsLastTokenOfNode<PatternSyntax>), but that check will fail if the keyword is already partially
+            // written, e.g. 'e is int o'. We thus need to include this special check.
+            if (patternSyntax != null && patternSyntax.IsKind(SyntaxKind.DeclarationPattern))
+            {
+                return true;
+            }
+
+            // e is { P: (1) $$
+            if (leftToken.IsKind(SyntaxKind.CloseParenToken) && node.IsParentKind(SyntaxKind.RecursivePattern))
+            {
+                return true;
+            }
+
+            // Getting rid of the extra parentheses to deal with cases such as 'case (((1 $$'
+            while (node.IsParentKind(SyntaxKind.ParenthesizedExpression) || node.IsParentKind(SyntaxKind.ParenthesizedPattern))
+            {
+                node = node.Parent;
+            }
```

Comment:
highly skptical this loop is needed (at least for the ParenthesizedPattern case)